src/main/scala/chisel3/util/circt/SizeOf.scala

```diff
@@ -0,0 +1,32 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.util.circt
+
+import scala.util.hashing.MurmurHash3
+
+import chisel3._
+import chisel3.experimental.{annotate, ChiselAnnotation, ExtModule}
+
+import circt.Intrinsic
+
+/** Create a module with a parameterized type which returns the size of the type
+  * as a compile-time constant.  This lets you write code which depends on the
+  * results of type inference.
+  */
+private class SizeOfIntrinsic[T <: Data](gen: T) extends ExtModule {
+  val i = IO(Input(gen));
+  val size = IO(Output(UInt(32.W)));
+  annotate(new ChiselAnnotation {
+    def toFirrtl =
+      Intrinsic(toTarget, "circt.sizeof")
+  })
+  override val desiredName = "SizeOf_" + MurmurHash3.stringHash(gen.toString()).toHexString
```

Comment:
I'm still slightly worried about this hash approach, especially given that `MurmurHash3` could realistically have collisions. I think this is a reasonable place to start with this, though.

Reply:
My concern is that the type's string won't always be well formed, but I don't know if that is justified.  If it is, then it could just be suitably escaped.  If it's not, then a hash doesn't help.  Hashing just seemed easier than escaping the string to be firrtl compatible.

Reply:
🤔 The reasoning makes sense. It may be problematic to use the type as this is only guaranteed to be unique for ground types (with the restriction that external modules cannot have uninformed widths). E.g., could a user could cause a collision if they have two bundles with the same name?

Reply:
Using the global object avoids the problem.